Defer setting msg_type = INBOX until message is handled

rowanseymour opened the pull request.

nicpottier commented on temba/msgs/models.py:
```diff
@@ -702,11 +703,18 @@ def mark_handled(cls, msg):
+        if not msg.msg_type:
+            msg.msg_type = INBOX
+            update_fields.append('msg_type')
+
```
Hrmm.. ya, this is where I don't love the addition of null, in that we now have two states that are undefined.. (INBOX && QUEUED and msg_type of None and HANDLED)

Having a single place where the state of the message is tracked seems a lot cleaner long term, as we have now.  Conflating that with the msg_type smells a bit to me.

Does the reduction of the where make any performance difference given our indexes now?

rowanseymour replied:
When a message is received and status is PENDING, if you think of INBOX as a "label" then it smells a lot more to apply that label by default, and then remove it when you've later realised it wasn't appropriate.

I don't think this change means that `msg_type` is tracking state anymore than it currently is. It's more that we explicitly use null instead of "this might be wrong so filter it out by `status!=PENDING`"

I'm not sure about performance as the motivating reason here was complexity rather than performance.

nicpottier replied:
Oh, there cases where we remove it when we realize it isn't appropriate? Are you just talking about archiving or something else? If we change our mind later then ya, that's a bigger argument for this for sure.

I definitely do see the point that waiting to mark the message type until it is truly handled makes sense, just the funky None state feels slightly odd to me.

I guess I'm fine with this as ya, I can see how it is going to make the triggers a lot easier to reason about. Just hate ambiguous states in the data model.

rowanseymour replied:
Let's rewind... currently when any non-IVR incoming message is created, it gets `msg_type=INBOX` and `status=PENDING`. Later if it's handled by a flow, `msg_type` is changed to `FLOW`. When message handling is complete, the status becomes HANDLED.

It depends on how you think of inbox as to what that initial state of INBOX means for a flow message. As far as the Inbox view (think virtual label) is concerned, `msg_type` is potentially "wrong" until `status=HANDLED`.

My initial idea was that messages would start with UNDEFINED type.. but then I figured that's what we have NULL for.

nicpottier replied:
Ah right, the FLOW case is what I wasn't thinking of and ya, that's super goofy.

Ya, ya, ok, fine, totally the right thing.

ericnewcomer replied:
We chatted a bit about this one last night and came to the same conclusion after a very similar discussion. Conflating status and type here is what is confusing.. But would agree with Rowan that setting the type before we know what type it is is definitely wrong, so should allow null.

rowanseymour commented:
I removed a migration I had added to update existing messages in pending state to have `msg_type=None` but this could cause problems with multiple server instances in different states of update. So as is, user may briefly see some pending messages in their inbox after we update.
